File: src/data/courses/dsa/linked-lists/merge-sorting/merge-k-sorted/code/python/merge_k_sorted.py
# ============================================================
# Merge K Sorted Lists
# ============================================================
import heapq
class Solution:
    def mergeKLists(self, lists):
        heap = []
        for i, l in enumerate(lists):
            if l: heapq.heappush(heap, (l.val, i, l))
        dummy = tail = ListNode(0)
        while heap:
            val, i, node = heapq.heappop(heap)
            tail.next = node; tail = tail.next
            if node.next:
                heapq.heappush(heap, (node.next.val, i, node.next))
        return dummy.next

File: src/data/courses/dsa/linked-lists/merge-sorting/merge-k-sorted/code/java/MergeKSorted.java
// ============================================================
// Merge K Sorted Lists
// ============================================================
import java.util.*;
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>((a,b)->a.val-b.val);
        for (ListNode l : lists) if (l != null) pq.offer(l);
        ListNode dummy = new ListNode(0), tail = dummy;
        while (!pq.isEmpty()) {
            ListNode node = pq.poll();
            tail.next = node; tail = tail.next;
            if (node.next != null) pq.offer(node.next);
        }
        return dummy.next;
    }
}
